### PR TITLE
[feature](exchange) enable shared exchange sink buffer to reduce RPC concurrency.

### PR DESCRIPTION
**Pull request by Mryange**
### What problem does this PR solve?

In the past, each exchange sink had its own sink buffer.  
If the query concurrency is n, there would be n * n RPCs running concurrently 
in a typical shuffle scenario (each sender instance can send data to all downstream instances).  
Here, we introduce support for shared sink buffers.  
This does not reduce the total number of RPCs but can limit the number of concurrent RPCs.

### Release note

None

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [x] No need to test or manual test. Explain why:
        - [x] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [x] No.
    - [ ] Yes. 

- Does this need documentation?
    - [x] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

File: be/src/pipeline/exec/exchange_sink_buffer.cpp [hunk 1]
```diff
@@ -411,48 +413,25 @@ void ExchangeSinkBuffer::_ended(InstanceLoId id) {
...
 }
 
 void ExchangeSinkBuffer::_set_receiver_eof(InstanceLoId id) {
     std::unique_lock<std::mutex> lock(*_instance_to_package_queue_mutex[id]);
     _instance_to_receiver_eof[id] = true;
+    // When the receiving side reaches eof, it means the receiver has finished early.
+    // The remaining data in the current rpc_channel does not need to be sent,
+    // and the rpc_channel should be turned off immediately.
     _turn_off_channel(id, lock);
-    std::queue<BroadcastTransmitInfo, std::list<BroadcastTransmitInfo>>& broadcast_q =
```
Comment (Gabriel39) [inline be/src/pipeline/exec/exchange_sink_buffer.cpp:427]: Keep this code block

File: be/src/pipeline/exec/exchange_sink_buffer.h [hunk 2]
```diff
@@ -214,6 +250,10 @@ class ExchangeSinkBuffer final : public HasTaskExecutionCtx {
     // One channel is corresponding to a downstream instance.
     phmap::flat_hash_map<InstanceLoId, bool> _rpc_channel_is_idle;
 
+    // There could be multiple situations that cause an rpc_channel to be turned off,
+    // such as receiving the eof, manual cancellation by the user, or all sinks reaching eos.
+    // Therefore, it is necessary to prevent an rpc_channel from being turned off multiple times.
+    phmap::flat_hash_map<InstanceLoId, bool> _rpc_channel_is_turn_off;
```
Comment (Gabriel39) [inline be/src/pipeline/exec/exchange_sink_buffer.h:256]: Could we unify these 2 maps?

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

File: be/src/runtime/runtime_state.h [hunk 3]
```diff
@@ -317,6 +317,12 @@ class RuntimeState {
 
     int per_fragment_instance_idx() const { return _per_fragment_instance_idx; }
 
+    const pipeline::PipelineFragmentContext* fragment_context() const { return _fragment_context; }
```
Comment (Gabriel39) [inline be/src/runtime/runtime_state.h:320]: Get this from pipeline task

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run beut

**Review by Gabriel39 — DISMISSED** [review]

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

File: be/src/pipeline/exec/exchange_sink_operator.cpp [hunk 4]
```diff
@@ -724,4 +741,35 @@ DataDistribution ExchangeSinkOperatorX::required_data_distribution() const {
...
+}
+
+/// TODO: Modify this to let FE handle the judgment instead of BE.
+std::shared_ptr<ExchangeSinkBuffer> ExchangeSinkOperatorX::get_sink_buffer(
+        InstanceLoId sender_ins_id) {
+    if (!_child) {
+        throw doris::Exception(ErrorCode::INTERNAL_ERROR,
+                               "ExchangeSinkOperatorX did not correctly set the child.");
+    }
+    if (std::dynamic_pointer_cast<SortSourceOperatorX>(_child)) {
```
Comment (HappenLee) [inline be/src/pipeline/exec/exchange_sink_operator.cpp:764]: better `if (_state->enable_shared_exchange_sink_buffer()
&& ! std::dynamic_pointer_cast<LocalExchangeSourceOperatorX>(_child)) && !std::dynamic_pointer_cast<SortSourceOperatorX>(_child))`

And add enough comment for the code

File: be/src/pipeline/exec/exchange_sink_buffer.h [hunk 5]
```diff
@@ -226,32 +278,43 @@ class ExchangeSinkBuffer final : public HasTaskExecutionCtx {
...
 
     std::atomic<int> _total_queue_size = 0;
-    std::shared_ptr<Dependency> _queue_dependency = nullptr;
-    std::shared_ptr<Dependency> _finish_dependency = nullptr;
-    std::shared_ptr<Dependency> _broadcast_dependency = nullptr;
-    ExchangeSinkLocalState* _parent = nullptr;
+
+    // _running_sink_count is used to track how many sinks have not finished yet.
+    // It is only decremented when eos is reached.
+    phmap::flat_hash_map<InstanceLoId, int64_t> _running_sink_count;
```
Comment (HappenLee) [inline be/src/pipeline/exec/exchange_sink_buffer.h:312]: rethink should we union the map. seems all map same key?

File: be/src/pipeline/exec/exchange_sink_buffer.cpp [hunk 6]
```diff
@@ -271,14 +285,16 @@ Status ExchangeSinkBuffer::_send_rpc(InstanceLoId id) {
             } else if (!s.ok()) {
                 _failed(id,
```
Comment (HappenLee) [inline be/src/pipeline/exec/exchange_sink_buffer.cpp:286]: should union the failed check？

**Comment by Mryange** [comment]
run buildall

**Review by Gabriel39 — APPROVED** [review]

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**